### PR TITLE
Initial 'Manage Organization' view

### PR DESCRIPTION
**Pull request by chenriksson**
Refactoring to reuse existing Account view for 'Manage Organization' (organization account)
- Moved common logic into base AccountsController class
  - Account view
  - Confirmation required
  - Change email, cancel change email
  - Change email subscriptions
- Introduced OrganizationsController for initial organization account logic
- Introduced ManageOrganization.cshtml view, based on Account.cshtml
- Moved common test logic into base AccountsControllerFacts class
- Updated Manage Organizations edit link to use the new view
- Updated transform confirmation to redirect to the new view

Add/Remove members will come in follow-up PR, as well as AAD tenant policy.

Screenshot:
*Note: using full 12 columns in order to support breadcrumb-style header*

image

### CODE REVIEW

File: src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+﻿@model ConfirmationViewModel
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml]: This appears, at first glance, to be nearly identical to `Users/ConfirmationRequired`. Can we share the code between the two more effectively and then change the strings based on whether or not the account whose email to confirm is an org or not?

File: src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            {
+                if (TempData.ContainsKey("ConfirmationRequiredMessage"))
+                {
+                    @ViewHelpers.AlertInfo(@<text>@((string)TempData["ConfirmationRequiredMessage"])</text>)
+                }
+
+                if (Model.AlreadyConfirmed)
+                {
+                    <p class="text-center">
+                        Your email address has already been confirmed.
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml]: `Your organization's email address` or something

File: src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                            <a href="@Url.OrganizationConfirmationRequired(account.Username)">email address has been verified.</a><br />
+                        </text>
+                    }
+                </text>)
+            }
+
+            @if (!account.Confirmed)
+            {
+                @ViewHelpers.AlertInfo(@<text>
+                    Uploading packages requires that you <a href="@Url.OrganizationConfirmationRequired(account.Username)">confirm the account</a>.
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml]: `confirm the organization`?

File: src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    <div class="row">
+        <div class="col-md-12">
+            <h1>
+                <a href="@Url.ManageMyOrganizations()">Manage Organizations</a>
+                &gt;
+                <a href="@Url.User(Model.Account)">@Model.AccountName</a>
+            </h1>
+        </div>
+
+        <div class="@ViewHelpers.GetColumnClasses(ViewBag)">
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml]: Can we move the parts of the page that are shared between users and organizations to their own partials?

File: src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                        @Html.HiddenFor(m => m.AccountName);
+
+                        <div class="checkbox">
+                            <label>
+                                @Html.ShowCheckboxFor(x => x.ChangeNotifications.EmailAllowed)
+                                <b>Users can contact this organization through the @Config.Current.Brand</b>
+                            </label>
+                            <div class="label-sibling">
+                                This setting allows registered users of the site to contact this organization about
+                                packages that the organization owns using the <em>Contact Owners</em> form, or to request
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml]: nit: `contact this organization about packages that the organization owns` -> `contact this organization about packages that it owns`

File: tests/NuGetGallery.Facts/Controllers/AccountsControllerFacts.cs [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                return GetController<TAccountsController>();
+            }
+
+            protected TUser GetAccount(TAccountsController controller)
+            {
+                var environment = controller.OwinContext.Environment;
+                if (!environment.TryGetValue(AccountEnvironmentKey, out var account) || !(account is TUser))
+                {
+                    account = Activator.CreateInstance<TUser>();
+                    ((TUser)account).Key = 2345;
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Controllers/AccountsControllerFacts.cs]: `var user = account as TUser;` so you don't have to keep casting?

File: tests/NuGetGallery.Facts/Controllers/OrganizationsControllerFacts.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+using System.Threading.Tasks;
+using System.Web.Mvc;
+using Xunit;
+
+namespace NuGetGallery
+{
+    public class OrganizationsControllerFacts
+        : AccountsControllerFacts<OrganizationsController, Organization, OrganizationAccountViewModel>
+    {
+        public static User GetAdminAsCurrentUser(OrganizationsController controller, Organization account)
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Controllers/OrganizationsControllerFacts.cs]: May want to specify it's an `OrganizationAdmin` and not a `SiteAdmin` with your naming.

File: tests/NuGetGallery.Facts/Controllers/OrganizationsControllerFacts.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        {
+            if (controller.GetCurrentUser() == null)
+            {
+                var admin = CreateAdmin(account);
+                controller.SetCurrentUser(admin);
+            }
+            return controller.GetCurrentUser();
+        }
+
+        public static User CreateAdmin(Organization account)
```
Comment (scottbommarito) [inline tests/NuGetGallery.Facts/Controllers/OrganizationsControllerFacts.cs]: `TestUtility.FakeOrganizationAdmin` and `Fakes.OrganizationAdmin` are both good alternatives for creating your own users for testing.

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 9]
```diff
@@ -44,60 +39,44 @@ public partial class UsersController
...
 
-                model = new ConfirmationViewModel(user)
-                {
-                    SentEmail = true
-                };
-            }
-            else
+        protected override User GetAccount(string accountName)
+        {
+            var currentUser = GetCurrentUser();
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: UsersController already uses `GetCurrentUser` to get the account.

I'm just introducing `GetAccount(string accountName)` as an abstraction so that OrganizationsController can operate on a different account.

File: src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml (hunk 1 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml]: It is very identical; I went with a copy since there were so many inline messages that needed updating. I'll see if there's a good way I can re-use the code.

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGetGallery/Controllers/UsersController.cs]: Yes, but now that all of the methods in `UsersController` support passing in an account name, it's confusing that this method behaves as it does--restricting the current user from acting upon an account it may have permission to manage the account of.

File: src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml (hunk 1 above)
Comment (scottbommarito) [reply src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml]: add an `IsOrganization` field to the model

```
<p>
    @(Model.IsOrganization ? "Your organization" : "Your account") is confirmed.
</p>
```

File: src/NuGetGallery/Controllers/AccountsController.cs [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+using System.Linq;
+using System.Net;
+using System.Net.Mail;
+using System.Threading.Tasks;
+using System.Web.Mvc;
+using NuGetGallery.Authentication;
+
+namespace NuGetGallery
+{
+    public abstract class AccountsController<TUser, TAccountViewModel> : AppController
```
Comment (chenriksson) [inline src/NuGetGallery/Controllers/AccountsController.cs]: I recommend comparing against [UsersController](https://github.com/NuGet/NuGetGallery/blob/dev/src/NuGetGallery/Controllers/UsersController.cs), where most of this code was extracted from.

File: src/NuGetGallery/RouteNames.cs [hunk 11]
```diff
@@ -5,11 +5,9 @@ namespace NuGetGallery
 {
     public static class RouteName
     {
-        public const string V2ApiCuratedFeed = "V2ApiCuratedFeed";
-        public const string V1ApiFeed = "V1ApiFeed";
-        public const string V2ApiFeed = "V2ApiFeed";
```
Comment (skofman1) [inline src/NuGetGallery/RouteNames.cs]: Why were those removed? We don't use them?

File: src/NuGetGallery/Controllers/OrganizationsController.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+            EmailUpdateCancelled = Strings.OrganizationEmailUpdateCancelled,
+            EmailUpdated = Strings.OrganizationEmailUpdated,
+            EmailUpdatedWithConfirmationRequired = Strings.OrganizationEmailUpdatedWithConfirmationRequired
+        };
+
+        [HttpGet]
+        [Authorize]
+        public virtual ActionResult ManageOrganization(string accountName)
+        {
+            var account = UserService.FindByUsername(accountName) as Organization;
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/OrganizationsController.cs]: what happens if account is null because accountName is not an organization? Will we fail gracefully?

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: Parameter `accountName` is not required for the `UsersController`, for instance when used from the `Account()` action.

I can change to only use `GetCurrentUser` when accountName is not provided.

File: src/NuGetGallery/ViewModels/OrganizationMemberViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGetGallery
+{
+    public class OrganizationMemberViewModel
+    {
+        public OrganizationMemberViewModel(Membership membership)
+        {
+            Member = membership.Member;
```
Comment (skofman1) [inline src/NuGetGallery/ViewModels/OrganizationMemberViewModel.cs]: null checks

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGetGallery/Controllers/UsersController.cs]: Exactly, which is confusing--these actions have this parameter that's unused and in fact throws if there's anything provided. It's not that bad from a user perspective given that the routing configuration will prevent anybody from hitting these endpoints with unexpected users but having it set up like this makes the code misleading.

File: src/NuGetGallery/RouteNames.cs (hunk 11 above)
Comment (chenriksson) [reply src/NuGetGallery/RouteNames.cs]: Yes, these are not referenced anywhere.

File: src/NuGetGallery/Controllers/OrganizationsController.cs (hunk 12 above)
Comment (chenriksson) [reply src/NuGetGallery/Controllers/OrganizationsController.cs]: Changed call to `GetAccount(accountName)`. Action methods check for null and insufficient permissions and return a `Forbidden` status code for either case.

**Comment by chenriksson** [comment]
@scottbommarito @skofman1 - initial feedback has been addressed. I've also split the Account.cshtml shared code into partials for better code reuse.

File: src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml (hunk 4 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml]: done

File: src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml (hunk 1 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml]: done

File: src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml (hunk 2 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Organizations/ConfirmationRequired.cshtml]: done

File: src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml (hunk 5 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml]: done

File: tests/NuGetGallery.Facts/Controllers/AccountsControllerFacts.cs (hunk 6 above)
Comment (chenriksson) [reply tests/NuGetGallery.Facts/Controllers/AccountsControllerFacts.cs]: done

File: tests/NuGetGallery.Facts/Controllers/OrganizationsControllerFacts.cs (hunk 7 above)
Comment (chenriksson) [reply tests/NuGetGallery.Facts/Controllers/OrganizationsControllerFacts.cs]: done

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: Changed to return null if accountName provided but does not match. Action methods using `GetAccount` will return forbidden if this is the case. Leaving accountName as optional for UsersController, just in case.

File: src/NuGetGallery/Services/ActionsRequiringPermissions.cs [hunk 14]
```diff
@@ -86,5 +86,12 @@ public static class ActionsRequiringPermissions
         public static ActionRequiringAccountPermissions HandlePackageOwnershipRequest = 
             new ActionRequiringAccountPermissions(
                 accountPermissionsRequirement: RequireOwnerOrOrganizationAdmin);
+
+        /// <summary>
+        /// The action of managing an account.
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Services/ActionsRequiringPermissions.cs]: done

File: src/NuGetGallery/ViewModels/OrganizationMemberViewModel.cs (hunk 13 above)
Comment (chenriksson) [reply src/NuGetGallery/ViewModels/OrganizationMemberViewModel.cs]: done.

File: src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml (hunk 3 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Organizations/ManageOrganization.cshtml]: done

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/NuGetGallery/Controllers/UsersController.cs]: I like this a lot better, thanks. Might still want to include a comment specifying why we are behaving like this.

File: src/NuGetGallery/Views/Shared/_AccountChangeNotifications.cshtml [hunk 15]
```diff
@@ -0,0 +1,88 @@
+﻿@model AccountViewModel
+@{
+    var controllerName = Model.IsOrganization ? "Organizations" : "Users";
+
+    ViewBag.NotifyPackagePushedEnabled = "Receiving package push notifications";
+    ViewBag.NotifyPackagePushedNotEnabled = "Not receiving package push notifications";
+
+    if (Model.IsOrganization)
+    {
+        ViewBag.EmailAllowedEnabled = "Users can contact";
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Shared/_AccountChangeNotifications.cshtml]: the strings below are `Users can contact you`
should this be `Users can contact the/your organization`?

File: tests/NuGetGallery.Facts/Controllers/OrganizationsControllerFacts.cs (hunk 8 above)
Comment (chenriksson) [reply tests/NuGetGallery.Facts/Controllers/OrganizationsControllerFacts.cs]: done

File: src/NuGetGallery/Views/Shared/_AccountChangeNotifications.cshtml (hunk 15 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Shared/_AccountChangeNotifications.cshtml]: Real estate is really limited here, so I chose to shorten instead:

image

Organization page would show:
> Users can [not] contact | [Not] Receiving package push notifications

File: src/NuGetGallery/Views/Shared/_ConfirmationResendForm.cshtml [hunk 16]
```diff
@@ -1,11 +1,14 @@
...
 {
     @Html.AntiForgeryToken()
 
     <div class="form-group">
         <input type="submit" class="btn btn-primary form-control" value="Resend Confirmation Email" />
     </div>
 }
-<p class="text-center">Is @Model.UnconfirmedEmailAddress incorrect? <span class="ms-noWrap">@Html.ActionLink("Change your email", "Account", "Users")</span></p>
\ No newline at end of file
+<p class="text-center">Is @Model.UnconfirmedEmailAddress incorrect? <span class="ms-noWrap">@Html.ActionLink("Change your email", "ManageOrganization", "Organizations")</span></p>
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Shared/_ConfirmationResendForm.cshtml]: Why is this org specific? Should it use `controllerName`?

File: tests/NuGetGallery.Facts/Controllers/AccountsControllerFacts.cs [hunk 17]
```diff
@@ -0,0 +1,695 @@
...
+using System.Net.Mail;
+using System.Threading.Tasks;
+using System.Web.Mvc;
+using Moq;
+using NuGetGallery.Framework;
+using Xunit;
+
+namespace NuGetGallery
+{
+    public class AccountsControllerFacts<TAccountsController, TUser, TAccountViewModel>
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/Controllers/AccountsControllerFacts.cs]: So we run all of these facts twice, once for organizations, once for users?

File: src/NuGetGallery/Views/Shared/_ConfirmationResendForm.cshtml (hunk 16 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Shared/_ConfirmationResendForm.cshtml]: Ooops! Missed that - thanks!

File: tests/NuGetGallery.Facts/Controllers/AccountsControllerFacts.cs (hunk 17 above)
Comment (chenriksson) [reply tests/NuGetGallery.Facts/Controllers/AccountsControllerFacts.cs]: Yes, `AccountsControllerFacts` tests are inherited and run under both contexts:
- `OrganizationsController`
  - currentUser = `Fakes.User`
  - account = `Fakes.User`
- `OrganizationsController`
  - currentUser = `Fakes.OrganizationAdmin`
  - account = `Fakes.Organization`

**Comment by joelverhagen** [comment]
👏 nice job on the refactoring @chenriksson!
